File: android-reference-app/app/src/main/java/io/agora/android_reference_app/CallQualityActivity.java
package io.agora.android_reference_app;

import io.agora.agora_manager.AgoraManager;
import io.agora.call_quality_manager.CallQualityManager;
import io.agora.rtc2.IRtcEngineEventHandler;

import android.graphics.Color;
import android.os.Bundle;
import android.view.Gravity;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.TextView;

public class CallQualityActivity extends BasicImplementationActivity {
    private CallQualityManager callQualityManager;

    private TextView networkStatus; // For updating the network status
    private boolean isEchoTestRunning = false; // Keeps track of the echo test
    private Button btnEchoTest;
    private final AgoraManager.AgoraManagerListener baseListener = getAgoraManagerListener();
    private TextView overlayText;

    private void updateNetworkStatus(int quality) {
        if (quality > 0 && quality < 3) networkStatus.setBackgroundColor(Color.GREEN);
        else if (quality <= 4) networkStatus.setBackgroundColor(Color.YELLOW);
        else if (quality <= 6) networkStatus.setBackgroundColor(Color.RED);
        else networkStatus.setBackgroundColor(Color.WHITE);

        networkStatus.setText(Integer.toString(quality));
    }

    @Override
    protected int getLayoutResourceId() {
        return R.layout.activity_call_quality;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        networkStatus = findViewById(R.id.networkStatus);
        btnEchoTest = findViewById(R.id.btnEchoTest);

        // Start the probe test
        callQualityManager.startProbeTest();
    }

    @Override
    protected void initializeAgoraManager() {
        callQualityManager = new CallQualityManager(this);
        agoraManager = callQualityManager;

        // Set the current product depending on your application
        agoraManager.setCurrentProduct(AgoraManager.ProductName.VIDEO_CALLING);
        // Set up a listener for updating the UI
        agoraManager.setListener(new CallQualityManager.CallQualityManagerListener() {

            @Override
            public void onNetworkQuality(int uid, int txQuality, int rxQuality) {
                // Use down-link network quality to update the network status
                runOnUiThread(() -> updateNetworkStatus(rxQuality));
            }

            @Override
            public void onLastMileQuality(int quality) {
                runOnUiThread(() -> updateNetworkStatus(quality));
            }

            @Override
            public void onUserJoined(int uid) {
                if (overlayText == null ) setupOverlayText();
            }

            @Override
            public void onRemoteVideoStats(IRtcEngineEventHandler.RemoteVideoStats stats) {
                int selectedUserId = (int) mainFrame.getTag();
                if (selectedUserId == agoraManager.localUid) {
                    runOnUiThread(() -> overlayText.setText(""));
                    return;
                }
                if (selectedUserId == stats.uid)  {
                    String caption = "Uid: " + stats.uid
                            + "\nRenderer frame rate: " + stats.rendererOutputFrameRate
                            + "\nReceived bitrate: " + stats.receivedBitrate
                            + "\nPublish duration: " + stats.publishDuration
                            + "\nFrame loss rate: " + stats.frameLossRate;
                    runOnUiThread(() -> overlayText.setText(caption)
                    );
                }
            }

            @Override
            public void onMessageReceived(String message) {
                baseListener.onMessageReceived(message);
            }

            @Override
            public void onRemoteUserJoined(int remoteUid, SurfaceView surfaceView) {
                baseListener.onRemoteUserJoined(remoteUid, surfaceView);
                // Choose low quality when the video plays in a small frame
                callQualityManager.setStreamQuality(remoteUid, false);
            }

            @Override
            public void onRemoteUserLeft(int remoteUid) {
                runOnUiThread(() -> overlayText.setText(""));
                baseListener.onRemoteUserLeft(remoteUid);
            }
        });
    }

    @Override
    protected void join() {
        int result = callQualityManager.joinChannelWithToken();
        if (result == 0) {
            // Start local video
            showLocalVideo();
            btnJoinLeave.setText(R.string.leave);
            btnEchoTest.setEnabled(false);
            if (radioGroup.getVisibility() != View.GONE) radioGroup.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    protected void leave() {
        super.leave();
        btnEchoTest.setEnabled(true);
        overlayText = null;
    }

    public void echoTest(View view) {
        if (!isEchoTestRunning) {
            btnEchoTest.setText(R.string.stop_echo_test);
            surfaceViewMain = callQualityManager.startEchoTest();
            mainFrame.addView(surfaceViewMain);
            isEchoTestRunning = true;
        } else {
            callQualityManager.stopEchoTest();
            btnEchoTest.setText(R.string.start_echo_test);
            mainFrame.removeView(surfaceViewMain);
            isEchoTestRunning = false;
        }
        btnJoinLeave.setEnabled(!isEchoTestRunning);
    }

    public void setupOverlayText() {
        // Create a new TextView
        overlayText = new TextView(this);
        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT
        );
        layoutParams.gravity = Gravity.BOTTOM; // Set gravity to bottom left
        overlayText.setLayoutParams(layoutParams);
        overlayText.setTextSize(14);
        overlayText.setTextColor(Color.WHITE);
        overlayText.setPadding(10, 0, 0, 0);
        // Add the TextView to the FrameLayout
        runOnUiThread(() ->
                mainFrame.addView(overlayText));
    }

    @Override
    protected void swapVideo(int frameId) {
        // Switch to high-quality for the remote video going into the main frame
        FrameLayout smallFrame = findViewById(frameId);
        int smallFrameUid = (int) smallFrame.getTag();
        if (smallFrameUid != agoraManager.localUid)
            callQualityManager.setStreamQuality(smallFrameUid, true);

        // Switch to low-quality for the remote video going into the small frame
        int mainFrameUid = (int) mainFrame.getTag();
        if (mainFrameUid != agoraManager.localUid)
            callQualityManager.setStreamQuality(mainFrameUid, false);

        // Swap the videos
        super.swapVideo(frameId);
    }
}

File: android-reference-app/app/src/main/java/io/agora/android_reference_app/CallQualityActivity.kt
package io.agora.android_reference_app

import io.agora.rtc2.IRtcEngineEventHandler.RemoteVideoStats
import io.agora.call_quality_manager.CallQualityManager
import io.agora.call_quality_manager.CallQualityManager.CallQualityManagerListener

import android.graphics.Color
import android.widget.TextView
import android.os.Bundle
import android.view.SurfaceView
import android.widget.FrameLayout
import android.view.Gravity
import android.view.View
import android.widget.Button

class CallQualityActivity : BasicImplementationActivity() {
    private lateinit var callQualityManager: CallQualityManager
    private var btnEchoTest: Button? = null
    private var networkStatus: TextView? = null // For updating the network status
    private var overlayText: TextView? = null
    private var isEchoTestRunning = false // Keeps track of the echo test
    private val baseListener = agoraManagerListener
    
    // Override the UI layout
    override val layoutResourceId: Int
        get() = R.layout.activity_call_quality

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Set up access to the UI elements
        networkStatus = findViewById(R.id.networkStatus)
        btnEchoTest = findViewById(R.id.btnEchoTest)
    }

    override fun initializeAgoraManager() {
        // Instantiate an object of the callQualityManager class, which is an extension of the AgoraManager
        callQualityManager = CallQualityManager(this)
        agoraManager = callQualityManager

        // Set up a listener for updating the UI
        agoraManager.setListener(object : CallQualityManagerListener {
            override fun onNetworkQuality(uid: Int, txQuality: Int, rxQuality: Int) {
                // Use down-link network quality to update the network status
                runOnUiThread { updateNetworkStatus(rxQuality) }
            }

            override fun onLastMileQuality(quality: Int) {
                runOnUiThread { updateNetworkStatus(quality) }
            }

            override fun onUserJoined(uid: Int) {
                if (overlayText == null) setupOverlayText()
            }

            override fun onRemoteVideoStats(stats: RemoteVideoStats?) {
                // Show remote video stats for the video in the main frame
                val selectedUserId = mainFrame.tag as Int
                if (selectedUserId == agoraManager.localUid) {
                    runOnUiThread { overlayText!!.text = "" }
                    return
                }
                if (selectedUserId == stats!!.uid) {
                    val caption = """
                        Uid: ${stats.uid}
                        Renderer frame rate: ${stats.rendererOutputFrameRate}
                        Received bitrate: ${stats.receivedBitrate}
                        Publish duration: ${stats.publishDuration}
                        Frame loss rate: ${stats.frameLossRate}
                        """.trimIndent()
                    runOnUiThread { overlayText!!.text = caption }
                }
            }

            override fun onMessageReceived(message: String?) {
                baseListener.onMessageReceived(message)
            }

            override fun onRemoteUserJoined(remoteUid: Int, surfaceView: SurfaceView?) {
                baseListener.onRemoteUserJoined(remoteUid, surfaceView)
                // Choose low quality when the video plays in a small frame
                callQualityManager.setStreamQuality(remoteUid, false)
            }

            override fun onRemoteUserLeft(remoteUid: Int) {
                // Hide the overlay text
                runOnUiThread { overlayText!!.text = "" }
                baseListener.onRemoteUserLeft(remoteUid)
            }

            override fun onJoinChannelSuccess(channel: String?, uid: Int, elapsed: Int) {
                baseListener.onJoinChannelSuccess(channel, uid, elapsed)
                // Disable the echo test button
                btnEchoTest!!.isEnabled = false
            }
        })
    }

    override fun join() {
        callQualityManager.joinChannelWithToken()
    }

    override fun leave() {
        super.leave()
        btnEchoTest!!.isEnabled = true
        overlayText = null
    }

    fun echoTest(view: View) {
        if (!isEchoTestRunning) {
            btnEchoTest!!.setText(R.string.stop_echo_test)
            surfaceViewMain = callQualityManager.startEchoTest()
            mainFrame.addView(surfaceViewMain)
            isEchoTestRunning = true
        } else {
            callQualityManager.stopEchoTest()
            btnEchoTest!!.setText(R.string.start_echo_test)
            mainFrame.removeView(surfaceViewMain)
            isEchoTestRunning = false
        }
        btnJoinLeave.isEnabled = !isEchoTestRunning
    }

    fun setupOverlayText() {
        // Create a new TextView
        overlayText = TextView(this)
        val layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        )
        layoutParams.gravity = Gravity.BOTTOM // Set gravity to bottom left
        overlayText!!.layoutParams = layoutParams
        overlayText!!.textSize = 14f
        overlayText!!.setTextColor(Color.WHITE)
        overlayText!!.setPadding(10, 0, 0, 0)
        // Add the TextView to the FrameLayout
        runOnUiThread { mainFrame.addView(overlayText) }
    }

    override fun swapVideo(frameId: Int) {
        // Switch to high-quality for the remote video going into the main frame
        val smallFrame = findViewById<FrameLayout>(frameId)
        val smallFrameUid = smallFrame.tag as Int
        if (smallFrameUid != agoraManager.localUid) callQualityManager.setStreamQuality(
            smallFrameUid,
            true
        )

        // Switch to low-quality for the remote video going into the small frame
        val mainFrameUid = mainFrame.tag as Int
        if (mainFrameUid != agoraManager.localUid) callQualityManager.setStreamQuality(
            mainFrameUid,
            false
        )

        // Swap the videos
        super.swapVideo(frameId)
    }

    private fun updateNetworkStatus(quality: Int) {
        // Update the UI to indicate the network status
        when (quality) {
            in 1..2 -> networkStatus?.setBackgroundColor(Color.GREEN)
            in 3..4 -> networkStatus?.setBackgroundColor(Color.YELLOW)
            in 5..6 -> networkStatus?.setBackgroundColor(Color.RED)
            else -> networkStatus?.setBackgroundColor(Color.WHITE)
        }
        networkStatus!!.text = quality.toString()
    }
}